packages/alfa-style/src/property/contain.ts

```diff
@@ -0,0 +1,120 @@
+import {
+  Keyword,
+  Token,
+  Tuple,
+  type Parser as CSSParser,
+} from "@siteimprove/alfa-css";
+import { Parser } from "@siteimprove/alfa-parser";
+import { Slice } from "@siteimprove/alfa-slice";
+import { Err, Result } from "@siteimprove/alfa-result";
+
+import { Longhand } from "../longhand.js";
+
+const { either, map } = Parser;
+
+type Specified = Specified.Single | Specified.Multiple;
+type Computed = Specified;
+
+namespace Specified {
+  export type Single = Tuple<
+    [Keyword<"none"> | Keyword<"strict"> | Keyword<"content">]
+  >;
+  export type Multiple = Tuple<
+    [
+      Keyword<"size"> | Keyword<"inline-size">,
+      Keyword<"layout">,
+      Keyword<"style">,
+      Keyword<"paint">,
+    ]
+  >;
+}
+
+const parseSingle: CSSParser<Specified.Single> = map(
+  Keyword.parse("none", "strict", "content"),
+  (keyword) => Tuple.of(keyword),
+);
+const parseMultiple: CSSParser<Specified.Multiple> = (input) => {
+  let size: Keyword<"size"> | Keyword<"inline-size"> | undefined;
+  let layout: Keyword<"layout"> | undefined;
+  let style: Keyword<"style"> | undefined;
+  let paint: Keyword<"paint"> | undefined;
+
+  while (true) {
+    for ([input] of Token.parseWhitespace(input)) {
+    }
+
+    if (size === undefined) {
+      const result = Keyword.parse("size", "inline-size")(input);
+
+      if (result.isOk()) {
+        [input, size] = result.get();
+        continue;
+      }
+    }
+
+    if (layout === undefined) {
+      const result = Keyword.parse("layout")(input);
+
+      if (result.isOk()) {
+        [input, layout] = result.get();
+        continue;
+      }
+    }
+
+    if (style === undefined) {
+      const result = Keyword.parse("style")(input);
+
+      if (result.isOk()) {
+        [input, style] = result.get();
+        continue;
+      }
+    }
+
+    if (paint === undefined) {
+      const result = Keyword.parse("paint")(input);
+
+      if (result.isOk()) {
+        [input, paint] = result.get();
+        continue;
+      }
+    }
+
+    break;
+  }
+
+  if (
+    size === undefined &&
+    layout === undefined &&
+    style === undefined &&
+    paint === undefined
+  ) {
+    return Err.of("Expected at least one keyword");
+  }
+
+  return Result.of([
+    input,
+    Tuple.of(
+      ...[size, layout, style, paint].filter(
+        (keyword) => keyword !== undefined,
+      ),
+    ) as Specified.Multiple,
```

Comment:
**Issue:** This actually seems unsafe 🤔 `Tuple` have fixed length, so the type always expects 4 arguments but this may have as little as one after the filtering. Shouldn't this instead default to the default values and thus always create the correct tuple?
`Tuple.of([size ?? sizeDefault, …])`

See, e.g., `Shadow.parse` in `alfa-css`.